### PR TITLE
[WIP] Add PeerTube with OAuth integration service

### PR DESCRIPTION
**Pull request by tytan652**
#### Idea
**Note: This is Work In Progress, CMake right now disable by default the code about PeerTube**

**Another note: WIP commit about CMake exist because I'm anticipating #4560 merge, so I separate them to modify them more easily later**

[PeerTube](https://joinpeertube.org/en_US/#what-is-peertube) is a decentralized and free/libre video broadcasting service.

Since version 3, PeerTube has a live-stream functionality. Each stream has his own stream key, so it change for each user's live except if it's a permanent one (permanent live mode). More info [here](https://docs.joinpeertube.org/use-create-upload-video)

Due to it's decentralized nature, his OBS service can't be focused on one pre-established URL and RTMP default port.

The user can already use OBS "Custom..." service to stream to his PeerTube instance.

So I want to create an integration:
- where the user put his instance URL, make him able to connect with [OAuth](https://docs.joinpeertube.org/api-rest-reference.html) (and check if the instance exist and has live-stream enabled) to obtain automatically the RTMP server and stream key from a already created live or create a live from OBS.
- This service will only work with OAuth, because in my opinion "Custom..." is enough for the one who don't want to use OAuth.
- Get rid of the stream key once the live-stream is ended if it's non-permanent live.
- And maybe more integration thing in the future

Note: I want to use the term "instance" for PeerTube server and "server" for RTMP server on the UI and code.

#### Changes
- [x] Divide OAuth classes into OAuth and BrowserOAuth ones.
- [x] Add BasicOAuth classes.
- [x] Remove `#ifdef BROWSER_AVAILABLE` on some piece of code because auth is no longer browser-only.
- [x] Create a third type of rtmp-service "peertube", I don't know if I should make it less PeerTube centered.
- [x] Create PeerTube service integration login ability
- [ ] Add PeerTube service integration with already created live requester
- [ ] Add the ability to create a live from OBS
- [ ] Make OBS drop the streamkey (if non permanent) once the live ended.
- [ ] Make the new UIs look good

### Motivation and Context

I create this early PR to put on words my ideas for anyone asking what I am doing, and show you my work and get advice.

### Types of changes

- New feature (non-breaking change which adds functionality)

### Checklist:

- [ ] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [ ] The code has been tested.
- [ ] All commit messages are properly formatted and commits squashed where appropriate.
- [ ] I have included updates to all appropriate documentation.

### CODE REVIEW

File: UI/CMakeLists.txt [hunk 1]
```diff
@@ -170,6 +168,16 @@ if(BROWSER_AVAILABLE_INTERNAL)
 	endif()
 endif()
 
+if(PEERTUBE_ENABLED)
+	add_definitions(-DPEERTUBE_ENABLED)
```
Comment (PatTheMav) [inline UI/CMakeLists.txt:172]: Don't use `add_definitions`, as it adds compile flags to all(!) targets. I guess in your case you need to add a check to `rtmp-services` to check for the enable switch and then do 

```
    target_compile_definitions(rtmp-services
        PRIVATE PEERTUBE_ENABLED
```

and repeat the same in this file with the `obs` target.

**Comment by tytan652** [comment]
I write this message to make sure there is no misunderstanding about this *service*.

PeerTube is a self-hostable software that provide livestream capability, so PeerTube is not a centralized service like Twitch, YouTube so no pre-established central server.

This service will ask for an instance to the user. There is no default instance thing. So I don't know if asking for an approval by PeerTube developer is needed.
We could see it more like a stream only third-party client for PeerTube instances.

This PR will add new service a little like `Custom...` but specialized for only PeerTube instances with dedicated integrations.

**Comment by RytoEX** [comment]
I do wish that this could be achieved with less PeerTube-centric code and less code duplication.  It seems like this service is doing the same sort of flow that Dacast is using, but with a user-specified server URL/endpoint.  Maybe it's possible to create a generic OAuth flow and/or a generic "use OAuth to get stream key" flow that can be leveraged for things like this.

As an aside, I sort of wish that instead of maintaining this in our codebase, PeerTube could maintain and provide something like `auth-peertube.dll` as an OAuth plugin for the specific code needed for this.  I feel the same about the rest of the OAuth providers.

**Comment by tytan652** [comment]
> I do wish that this could be achieved with less PeerTube-centric code and less code duplication. It seems like this service is doing the same sort of flow that Dacast is using, but with a user-specified server URL/endpoint. Maybe it's possible to create a generic OAuth flow and/or a generic "use OAuth to get stream key" flow that can be leveraged for things like this.

If you consider the BrowserOAuth and BasicOAuth existence as code duplication, I did this because PeerTube does not need browser interaction so if OBS is build without browser PeerTube integration still there.
PeerTube instances expose the client id and secret publicly through API request.

I create BasicOAuth if one day another service doesn't need browser interaction.

This PR will not get just **a** streamkey, it will make the user if he has multiple livestreams on his account be able to choose which one or create a new one.

> As an aside, I sort of wish that instead of maintaining this in our codebase, PeerTube could maintain and provide something like `auth-peertube.dll` as an OAuth plugin for the specific code needed for this. I feel the same about the rest of the OAuth providers.

What is behind PeerTube is not a company but Framasoft a french association. Please don't consider they can afford easily the maintenance of this integration.
And me I'm just trying to create and maintain this integration voluntarily.

If the service API existed and was complete, maybe this PR could be rather a third-party plugin. But since it's not the case.

Maybe one day if service API is complete, separating all integrations from OBS as plugins will be needed.

**Comment by dodgepong** [comment]
So hang on, you're openly admitting that there are no plans by anyone associated with PeerTube to maintain this integration? I feel like that alone is grounds for closure of this PR.

**Comment by rogue-ronin** [comment]
"openly admitting"? You've been watching too much _Law & Order_ :smile:

**Comment by dodgepong** [comment]
Dunno what that comment is supposed to mean, but if this is a drive-by PR where the developers have no intention of maintaining it, then we're not going to be motivated to merge it. That's just the facts of open source.

**Comment by tytan652** [comment]
It's been a while, this is superseeded by:
- https://github.com/obsproject/rfcs/pull/45
- https://github.com/obsproject/rfcs/pull/39
- And a future plugin made feasible by those RFCs.

**Pull request closed without merging**